kernel-rs/src/virtio/virtio_disk.rs

```diff
@@ -352,6 +375,210 @@ impl VirtioDisk {
+        barray: &mut [Option<Buf>],
+        header: &mut Option<Descriptor>,
+        tailer: &mut Option<Descriptor>,
+        ctx: &KernelCtx<'_, '_>,
+    ) -> bool {
+        let mut start = 0usize;
+        let end = darray.len();
+        assert_eq!(end, barray.len());
+
+        // The header and tailer descriptors must have been allocated.
+        let hdesc = match header.take() {
+            Some(d) => d,
+            None => return false,
+        };
+
+        let tdesc = match tailer.take() {
+            Some(d) => d,
+            None => {
+                guard.get_pin_mut().as_mut().free(hdesc);
+                return false;
+            }
+        };
+
+        // Asserts that the chain contains at least one buffer to be written.
+        if end == 0 || darray[end - 1].is_none() {
+            guard.get_pin_mut().as_mut().free(hdesc);
+            guard.get_pin_mut().as_mut().free(tdesc);
+            return false;
+        }
+
+        while darray[start].is_none() {
+            start += 1;
+        }
+
+        let fbdesc = &darray[start].as_ref().unwrap();
+        let lbdesc = &darray[end - 1].as_ref().unwrap();
+
+        let mut this = guard.get_pin_mut().project();
+        let mut info = this.info.project();
+        let sector: usize = (barray[start].as_ref().unwrap()).blockno as usize * (BSIZE / 512);
+
+        let buf0 = &mut info.ops[hdesc.idx];
+        *buf0 = VirtIOBlockOutHeader::new(true, sector);
+
+        // Set the header and tailer descriptors and finalize the chain.
+        this.desc[hdesc.idx] = VirtqDesc {
+            addr: buf0 as *const _ as _,
+            len: mem::size_of::<VirtIOBlockOutHeader>() as _,
+            flags: VirtqDescFlags::NEXT,
+            next: fbdesc.idx as _,
+        };
+        info.inflight[hdesc.idx].status = true;
+        this.desc[tdesc.idx] = VirtqDesc {
+            addr: &info.inflight[hdesc.idx].status as *const _ as _,
+            len: 1,
+            flags: VirtqDescFlags::WRITE,
+            next: 0,
+        };
+        this.desc[lbdesc.idx].next = tdesc.idx as _;
+
+        // The request is complete when the last block is written.
+        // TODO: This may require that the `VIRTIO_F_IN_ORDER` feature has been negotiated.
+        info.inflight[hdesc.idx].b = &mut *barray[end - 1].as_mut().unwrap();
+
+        let ring_idx = this.avail.idx as usize % NUM;
+        this.avail.ring[ring_idx] = hdesc.idx as _;
+        fence(Ordering::SeqCst);
+
+        // Tell the device another avail ring entry is available.
+        this.avail.idx += 1;
+
+        fence(Ordering::SeqCst);
+
+        // SAFETY: all the three descriptors' fields are well set.
+        // Value is queue number.
+        unsafe {
+            MmioRegs::notify_queue(0);
+        }
+
+        // Wait for the disk to finishing writing the entire chain
+        (&barray[end - 1].as_ref().unwrap())
+            .vdisk_request_waitchannel
+            .sleep(guard, ctx);
+
```

Comment:
- Why is that the case? (about only the last `Buf` will get its field updated)
- I wait for only the last Buf's to wake up because it seems that only the last Buf will be notified upon completion because I tried waiting for `barray[start]` and then the program froze. (The specs state that, 
```
If (`VIRTIO_F_IN_ORDER`) negotiated, this knowledge allows devices to notify
the use of a batch of buffers to the driver by only writing out a single used ring entry with the id corresponding
to the head entry of the descriptor chain describing the last buffer in the batch.
```
so I am not sure about its correctness.)

Reply:
* At your code, you change the `disk` field to `true` for each `Buf` at [here](https://github.com/kaist-cp/rv6/pull/616/files#diff-2232b5da779bace84063c21a96a40942fbb5363d649c6087b822d808b1b64569R460).
* However, at [`intr`](https://github.com/kaist-cp/rv6/pull/616/files#diff-2232b5da779bace84063c21a96a40942fbb5363d649c6087b822d808b1b64569R610), we only change the `disk` field back to `false` only for a single `Buf`.

Reply:
Understood. I will fix that.